File: packages/sdk/src/webhook.ts
import { createHmac, timingSafeEqual } from 'node:crypto';

export const SIGNATURE_HEADER = 'X-ZettaPay-Signature';
export const TIMESTAMP_HEADER = 'X-ZettaPay-Timestamp';
export const EVENT_ID_HEADER = 'X-ZettaPay-Event-Id';
export const ATTEMPT_HEADER = 'X-ZettaPay-Attempt';

const SIGNATURE_PREFIX = 'sha256=';
const HEX_RE = /^[0-9a-f]+$/i;
const DEFAULT_TOLERANCE_SEC = 300;

export type WebhookFailureReason =
  | 'missing_event_id'
  | 'missing_signature'
  | 'malformed_signature'
  | 'missing_timestamp'
  | 'invalid_timestamp'
  | 'timestamp_out_of_tolerance'
  | 'signature_mismatch'
  | 'invalid_payload';

export type HeaderBag =
  | Record<string, string | string[] | undefined>
  | { get(name: string): string | null };

export interface ParseWebhookOptions<TPayload = unknown> {
  /** Merchant webhook secret. Must match the secret stored on the merchant record. */
  secret: string;
  /** Raw request body — `Buffer`, `Uint8Array`, or pre-decoded `string`. JSON re-encoding breaks the HMAC. */
  body: Buffer | Uint8Array | string;
  /** Headers from the inbound request. Accepts plain objects, Node `IncomingHttpHeaders`, or `Headers`. */
  headers: HeaderBag;
  /** Reject events whose timestamp drifts more than this many seconds from `now`. Defaults to 300s. */
  toleranceSec?: number;
  /** Override the clock for tests. */
  now?: () => number;
  /** Optional payload validator/parser. Receives the JSON-decoded body and may narrow the type. */
  parsePayload?: (raw: unknown) => TPayload;
}

export interface ParsedWebhook<TPayload = unknown> {
  /** Stable across retries — use this as your idempotency dedup key. */
  eventId: string;
  /** Epoch milliseconds emitted by the dispatcher (`X-ZettaPay-Timestamp`). */
  timestamp: number;
  /** 1-indexed attempt number from `X-ZettaPay-Attempt`. `null` if header is absent. */
  attempt: number | null;
  /** JSON-decoded body, optionally narrowed by `parsePayload`. */
  payload: TPayload;
  /** Verbatim raw body as a UTF-8 string. */
  rawBody: string;
}

export type ParseWebhookResult<TPayload = unknown> =
  | { valid: true; event: ParsedWebhook<TPayload> }
  | { valid: false; reason: WebhookFailureReason };

/**
 * Verifies a ZettaPay webhook (HMAC + timestamp drift) and extracts the dedup
 * key in one call. Returns `{ valid: true, event }` with the stable
 * `event.eventId` from `X-ZettaPay-Event-Id` so a merchant can dedupe before
 * processing — retries reuse the same id.
 *
 * Pass the *raw* request body (e.g. via `express.raw()`); re-serializing JSON
 * changes byte order and invalidates the signature.
 */
export function parseWebhook<TPayload = unknown>(
  opts: ParseWebhookOptions<TPayload>,
): ParseWebhookResult<TPayload> {
  const eventId = readHeader(opts.headers, EVENT_ID_HEADER);
  if (!eventId) return { valid: false, reason: 'missing_event_id' };

  const signature = readHeader(opts.headers, SIGNATURE_HEADER);
  if (!signature) return { valid: false, reason: 'missing_signature' };

  const timestampRaw = readHeader(opts.headers, TIMESTAMP_HEADER);
  if (!timestampRaw) return { valid: false, reason: 'missing_timestamp' };

  const timestamp = Number(timestampRaw);
  if (!Number.isFinite(timestamp)) {
    return { valid: false, reason: 'invalid_timestamp' };
  }

  const tolerance = opts.toleranceSec ?? DEFAULT_TOLERANCE_SEC;
  const now = opts.now ? opts.now() : Date.now();
  if (Math.abs(now - timestamp) > tolerance * 1000) {
    return { valid: false, reason: 'timestamp_out_of_tolerance' };
  }

  const provided = parseSignature(signature);
  if (!provided) return { valid: false, reason: 'malformed_signature' };

  const rawBody = bodyToString(opts.body);
  const expected = createHmac('sha256', opts.secret)
    .update(`${timestampRaw}.${rawBody}`)
    .digest('hex');

  const a = Buffer.from(expected, 'hex');
  const b = Buffer.from(provided, 'hex');
  if (a.length !== b.length || !timingSafeEqual(a, b)) {
    return { valid: false, reason: 'signature_mismatch' };
  }

  let decoded: unknown;
  try {
    decoded = rawBody.length === 0 ? null : JSON.parse(rawBody);
  } catch {
    return { valid: false, reason: 'invalid_payload' };
  }

  let payload: TPayload;
  try {
    payload = opts.parsePayload ? opts.parsePayload(decoded) : (decoded as TPayload);
  } catch {
    return { valid: false, reason: 'invalid_payload' };
  }

  const attemptHeader = readHeader(opts.headers, ATTEMPT_HEADER);
  const attempt = attemptHeader === null ? null : Number(attemptHeader);

  return {
    valid: true,
    event: {
      eventId,
      timestamp,
      attempt: attempt !== null && Number.isFinite(attempt) ? attempt : null,
      payload,
      rawBody,
    },
  };
}

/**
 * Tiny store contract used by `dedupe()`. Plug in Redis, Postgres, or any
 * backing store that implements these two methods. Operations are expected to
 * be atomic per-eventId; treat repeated calls as idempotent.
 */
export interface EventStore {
  has(eventId: string): boolean | Promise<boolean>;
  add(eventId: string): void | Promise<void>;
}

/**
 * In-memory `EventStore` for development, tests, and single-process workers.
 * Holds a bounded set of recent event ids — not safe for multi-instance
 * deployments, but useful for local fixtures and small workers.
 *
 * Eviction is FIFO once `maxEntries` is reached.
 */
export class MemoryEventStore implements EventStore {
  private readonly maxEntries: number;
  private readonly seen = new Set<string>();

  constructor(opts: { maxEntries?: number } = {}) {
    this.maxEntries = opts.maxEntries ?? 10_000;
  }

  has(eventId: string): boolean {
    return this.seen.has(eventId);
  }

  add(eventId: string): void {
    if (this.seen.has(eventId)) return;
    if (this.seen.size >= this.maxEntries) {
      const oldest = this.seen.values().next().value;
      if (oldest !== undefined) this.seen.delete(oldest);
    }
    this.seen.add(eventId);
  }

  /** Test/operational helper — clears the entire dedup window. */
  clear(): void {
    this.seen.clear();
  }

  /** Number of event ids currently retained. */
  get size(): number {
    return this.seen.size;
  }
}

export interface DedupeResult {
  /** `true` if the eventId had not been seen before this call. */
  fresh: boolean;
  /** `true` if the eventId was already in the store — the merchant should ack 200 and skip work. */
  duplicate: boolean;
}

/**
 * Atomically check-and-record an eventId in the given `EventStore`. Returns
 * `{ duplicate: true }` if the event has already been processed so the
 * merchant can short-circuit with a 200 OK response.
 */
export async function dedupe(eventId: string, store: EventStore): Promise<DedupeResult> {
  const seen = await store.has(eventId);
  if (seen) return { fresh: false, duplicate: true };
  await store.add(eventId);
  return { fresh: true, duplicate: false };
}

function readHeader(headers: HeaderBag, name: string): string | null {
  if (typeof (headers as { get?: unknown }).get === 'function') {
    const value = (headers as { get(name: string): string | null }).get(name);
    return value && value.length > 0 ? value : null;
  }
  const bag = headers as Record<string, string | string[] | undefined>;
  const lower = name.toLowerCase();
  const direct = bag[name] ?? bag[lower];
  if (direct === undefined) return null;
  const value = Array.isArray(direct) ? direct[0] : direct;
  return value && value.length > 0 ? value : null;
}

function parseSignature(raw: string): string | null {
  const trimmed = raw.trim();
  const value = trimmed.startsWith(SIGNATURE_PREFIX)
    ? trimmed.slice(SIGNATURE_PREFIX.length)
    : trimmed;
  if (value.length === 0 || value.length % 2 !== 0) return null;
  if (!HEX_RE.test(value)) return null;
  return value.toLowerCase();
}

function bodyToString(body: Buffer | Uint8Array | string): string {
  if (typeof body === 'string') return body;
  if (Buffer.isBuffer(body)) return body.toString('utf8');
  return Buffer.from(body).toString('utf8');
}

File: packages/sdk-python/zettapay/webhook.py
"""Webhook signature verification for ZettaPay merchants.

Mirrors the TypeScript SDK helper at ``packages/sdk/src/webhook.ts``: HMAC-SHA256
over ``"{timestamp}.{raw_body}"`` with a timing-safe compare, a 300s drift
window on ``X-ZettaPay-Timestamp``, and the stable ``X-ZettaPay-Event-Id`` so a
merchant can dedupe retries before doing work.

Pass the *raw* request body — re-encoding the JSON changes byte order and
invalidates the signature.

Example::

    from zettapay.webhook import parse_webhook

    result = parse_webhook(
        secret=os.environ["ZETTAPAY_WEBHOOK_SECRET"],
        body=request.get_data(),     # bytes, not request.json
        headers=request.headers,
    )
    if not result.valid:
        return ("", 400)
    if seen_before(result.event.event_id):
        return ("", 200)
    process(result.event.payload)
"""

from __future__ import annotations

import hashlib
import hmac
import json
import re
import time
from dataclasses import dataclass
from typing import Any, Callable, Mapping, Optional, Union

SIGNATURE_HEADER = "X-ZettaPay-Signature"
TIMESTAMP_HEADER = "X-ZettaPay-Timestamp"
EVENT_ID_HEADER = "X-ZettaPay-Event-Id"
ATTEMPT_HEADER = "X-ZettaPay-Attempt"

_SIGNATURE_PREFIX = "sha256="
_HEX_RE = re.compile(r"^[0-9a-f]+$", re.IGNORECASE)
_DEFAULT_TOLERANCE_SEC = 300

# Discrete failure reasons — match on these strings if you need to branch
# (e.g. log signature_mismatch differently from timestamp_out_of_tolerance).
MISSING_EVENT_ID = "missing_event_id"
MISSING_SIGNATURE = "missing_signature"
MALFORMED_SIGNATURE = "malformed_signature"
MISSING_TIMESTAMP = "missing_timestamp"
INVALID_TIMESTAMP = "invalid_timestamp"
TIMESTAMP_OUT_OF_TOLERANCE = "timestamp_out_of_tolerance"
SIGNATURE_MISMATCH = "signature_mismatch"
INVALID_PAYLOAD = "invalid_payload"

WEBHOOK_FAILURE_REASONS = frozenset(
    {
        MISSING_EVENT_ID,
        MISSING_SIGNATURE,
        MALFORMED_SIGNATURE,
        MISSING_TIMESTAMP,
        INVALID_TIMESTAMP,
        TIMESTAMP_OUT_OF_TOLERANCE,
        SIGNATURE_MISMATCH,
        INVALID_PAYLOAD,
    }
)

Body = Union[bytes, bytearray, memoryview, str]


@dataclass(frozen=True)
class ParsedWebhook:
    """A verified webhook delivery ready for the merchant to process."""

    # Stable across retries — use this as the idempotency dedup key.
    event_id: str
    # Epoch milliseconds emitted by the dispatcher (``X-ZettaPay-Timestamp``).
    timestamp: int
    # 1-indexed attempt number from ``X-ZettaPay-Attempt`` if present.
    attempt: Optional[int]
    # JSON-decoded body, optionally narrowed by ``parse_payload``.
    payload: Any
    # Verbatim raw body as a UTF-8 string — useful for re-signing in tests.
    raw_body: str


@dataclass(frozen=True)
class ParseWebhookResult:
    """Discriminated result of :func:`parse_webhook`.

    When ``valid`` is ``True``, ``event`` is populated. When ``False``,
    ``reason`` carries one of the ``*_REASON`` constants exposed by this module.
    """

    valid: bool
    event: Optional[ParsedWebhook] = None
    reason: Optional[str] = None


def parse_webhook(
    *,
    secret: str,
    body: Body,
    headers: Any,
    tolerance_sec: int = _DEFAULT_TOLERANCE_SEC,
    now_ms: Optional[Callable[[], int]] = None,
    parse_payload: Optional[Callable[[Any], Any]] = None,
) -> ParseWebhookResult:
    """Verify a ZettaPay webhook and return the parsed event.

    Args:
        secret: Merchant webhook secret (matches the value on the merchant
            record). Treat as a credential.
        body: Raw request body — ``bytes``, ``bytearray``, ``memoryview`` or
            pre-decoded UTF-8 ``str``. Re-encoding JSON breaks the HMAC.
        headers: Inbound headers. Accepts any object with ``.get(name)`` or
            ``.items()`` — works with ``dict``, ``http.client.HTTPMessage``,
            ``werkzeug``/``flask`` headers, ``starlette``/``fastapi`` headers
            and ``requests.structures.CaseInsensitiveDict``.
        tolerance_sec: Reject events whose timestamp drifts more than this
            many seconds from ``now``. Defaults to 300s.
        now_ms: Override the clock for tests; should return epoch milliseconds.
        parse_payload: Optional validator/narrower applied to the decoded JSON
            body. Raise to mark the payload invalid.
    """
    event_id = _read_header(headers, EVENT_ID_HEADER)
    if not event_id:
        return ParseWebhookResult(valid=False, reason=MISSING_EVENT_ID)

    signature = _read_header(headers, SIGNATURE_HEADER)
    if not signature:
        return ParseWebhookResult(valid=False, reason=MISSING_SIGNATURE)

    timestamp_raw = _read_header(headers, TIMESTAMP_HEADER)
    if not timestamp_raw:
        return ParseWebhookResult(valid=False, reason=MISSING_TIMESTAMP)
    try:
        timestamp = int(timestamp_raw)
    except (TypeError, ValueError):
        return ParseWebhookResult(valid=False, reason=INVALID_TIMESTAMP)

    now = now_ms() if now_ms is not None else int(time.time() * 1000)
    if abs(now - timestamp) > tolerance_sec * 1000:
        return ParseWebhookResult(valid=False, reason=TIMESTAMP_OUT_OF_TOLERANCE)

    provided = _parse_signature(signature)
    if provided is None:
        return ParseWebhookResult(valid=False, reason=MALFORMED_SIGNATURE)

    raw_body = _body_to_string(body)
    signed_input = f"{timestamp_raw}.{raw_body}".encode("utf-8")
    expected = hmac.new(secret.encode("utf-8"), signed_input, hashlib.sha256).hexdigest()

    if not hmac.compare_digest(expected, provided):
        return ParseWebhookResult(valid=False, reason=SIGNATURE_MISMATCH)

    try:
        decoded = None if len(raw_body) == 0 else json.loads(raw_body)
    except ValueError:
        return ParseWebhookResult(valid=False, reason=INVALID_PAYLOAD)

    try:
        payload = parse_payload(decoded) if parse_payload is not None else decoded
    except Exception:
        return ParseWebhookResult(valid=False, reason=INVALID_PAYLOAD)

    attempt_raw = _read_header(headers, ATTEMPT_HEADER)
    attempt: Optional[int]
    if attempt_raw is None:
        attempt = None
    else:
        try:
            attempt = int(attempt_raw)
        except (TypeError, ValueError):
            attempt = None

    return ParseWebhookResult(
        valid=True,
        event=ParsedWebhook(
            event_id=event_id,
            timestamp=timestamp,
            attempt=attempt,
            payload=payload,
            raw_body=raw_body,
        ),
    )


def _read_header(headers: Any, name: str) -> Optional[str]:
    lower = name.lower()

    get = getattr(headers, "get", None)
    if callable(get):
        for variant in (name, lower):
            try:
                value = get(variant)
            except TypeError:
                value = None
            value = _normalize_header_value(value)
            if value:
                return value

    items = getattr(headers, "items", None)
    if callable(items):
        try:
            iterable = items()
        except TypeError:
            iterable = ()
        for key, raw in iterable:
            if isinstance(key, bytes):
                key = key.decode("latin-1", errors="replace")
            if isinstance(key, str) and key.lower() == lower:
                value = _normalize_header_value(raw)
                if value:
                    return value
    return None


def _normalize_header_value(value: Any) -> Optional[str]:
    if value is None:
        return None
    if isinstance(value, (list, tuple)):
        value = value[0] if value else None
        if value is None:
            return None
    if isinstance(value, bytes):
        value = value.decode("utf-8", errors="replace")
    if isinstance(value, str):
        return value if value else None
    return str(value)


def _parse_signature(raw: str) -> Optional[str]:
    trimmed = raw.strip()
    value = (
        trimmed[len(_SIGNATURE_PREFIX) :]
        if trimmed.startswith(_SIGNATURE_PREFIX)
        else trimmed
    )
    if not value or len(value) % 2 != 0:
        return None
    if not _HEX_RE.match(value):
        return None
    return value.lower()


def _body_to_string(body: Body) -> str:
    if isinstance(body, str):
        return body
    if isinstance(body, (bytes, bytearray)):
        return bytes(body).decode("utf-8")
    if isinstance(body, memoryview):
        return body.tobytes().decode("utf-8")
    raise TypeError(
        f"body must be bytes, bytearray, memoryview or str — got {type(body).__name__}"
    )


__all__ = [
    "ATTEMPT_HEADER",
    "EVENT_ID_HEADER",
    "INVALID_PAYLOAD",
    "INVALID_TIMESTAMP",
    "MALFORMED_SIGNATURE",
    "MISSING_EVENT_ID",
    "MISSING_SIGNATURE",
    "MISSING_TIMESTAMP",
    "ParsedWebhook",
    "ParseWebhookResult",
    "SIGNATURE_HEADER",
    "SIGNATURE_MISMATCH",
    "TIMESTAMP_HEADER",
    "TIMESTAMP_OUT_OF_TOLERANCE",
    "WEBHOOK_FAILURE_REASONS",
    "parse_webhook",
]
